dace: Add package and version 0.15.1

tbennun opened the pull request:
This PR adds the [DaCe](https://github.com/spcl/dace) framework as `py-dace`. It also updates `py-aenum` to a newer version that DaCe requires in order to work.

tbennun committed:
dace: Add package and version 0.14.2

tldahlgren reviewed (commented):
Confirmed the affected version sha256 for the two packages and the branch for py-dace.

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("python@3.7:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+    # From setup.py
+    depends_on("cmake@3.15.0:", when="@0.14.2:", type=("build", "run"))
```
Missing `py-scikit-build` dependency. Technically, dependency should be `py-cmake`, not `cmake`.

tbennun replied:
Not necessarily. The Python cmake dependency ensures the existence of CMake (I don't think py-cmake has any Python API). If an existing CMake is found, the `setup.py` script will not look for py-cmake: https://github.com/spcl/dace/blob/v0.14.2/setup.py#L35

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,47 @@
+        default=False,
+    )
+
+    depends_on("python@3.7:", when="@0.14.2:", type=("build", "run"))
```
```suggestion
    depends_on("python@3.6:3.10", when="@0.14.2:", type=("build", "run"))
```

tbennun replied:
Apologies, I was relying on this: https://spcldace.readthedocs.io/en/latest/setup/installation.html#installation

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("py-setuptools", type="build")
+    # From setup.py
+    depends_on("cmake@3.15.0:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-numpy@1.19:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-networkx@2.5:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-astunparse@1.6.3:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-sympy@1.10:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-pyyaml@5.1:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-ply@3.8:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-websockets@10.3:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-requests@2.24:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-aenum@3.1.11:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-flask@2.2.2:", when="@0.14.2:", type=("build", "run"))
+    depends_on("py-dill@0.3.6:", when="@0.14.2:", type=("build", "run"))
```
These dependencies don't appear to be correct, they should be:
```python
    cmake_requires = ['scikit-build', 'cmake']                                               
    install_requires=[                                                                 
          'numpy', 'networkx >= 2.5', 'astunparse', 'sympy<=1.9', 'pyyaml', 'ply', 'websockets', 'requests', 'flask',
          'aenum >= 3.1', 'dataclasses; python_version < "3.7"', 'dill', 'pyreadline;platform_system=="Windows"',
          'typing-compat; python_version < "3.8"'                                        
      ] + cmake_requires,       
```

tbennun replied:
Fixed in the current version.

tbennun committed:
Update dependencies and package version

tbennun commented:
@adamjstewart I fixed all the dependencies (and added ones that were on setup.py but not on Spack). Wherever there is a difference, I pointed it out with comments. This is ready for re-review on my end.

becker33 commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    version("master", branch="master", submodules=True)
+    version("0.15.1", sha256="69bfdbbd5c7177f2926a874f5fa82fcdef61fc532c022b4bc12e1e9218724093")
+
+    variant(
```
Are there no differences in the build for this variant?

tbennun replied:
at the very end a dependency on Likwid is added. The package detects if it is installed at runtime.

becker33 replied:
Yeah I saw the dependendency but I was worried about the lack of a build option.

adamjstewart replied:
Welcome to Python 😆

becker33 reviewed (commented):
One question, otherwise LGTM

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on("py-fparser@0.1.3:", type=("build", "run"))
+
+    depends_on("py-pyreadline", when="platform=win32", type=("build", "run"))
+    depends_on("py-dataclasses", when="^python@:3.6", type=("build", "run"))
+    depends_on("py-typing-compat", when="^python@:3.7", type=("build", "run"))
```
Spack no longer supports Python 3.7 and older, you could just remove these lines

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on("py-networkx@2.5:", type=("build", "run"))
+    depends_on("py-pyyaml", type=("build", "run"))
+    depends_on("py-ply", type=("build", "run"))
+    depends_on("py-websockets", type=("build", "run"))
```
Missing deps on requests, flask

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+
+    # Dependencies in setup.py with specific constraints in requirements.txt
+    depends_on("py-numpy@1.21:", type=("build", "run"))
+    depends_on("py-astunparse@1.6.3:", type=("build", "run"))
```
It would be much easier to review this PR if you keep all dependencies in the same order they are listed in setup.py

tbennun replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    # A typo in the 0.15.1 setup.py specifies sympy <= 1.9 instead of >= 1.9
+    depends_on("py-sympy@1.9:", type=("build", "run"))
+
+    # Dependencies in setup.py with specific constraints in requirements.txt
```
I don't normally trust requirements.txt since it isn't used by pip

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on("py-dataclasses", when="^python@:3.6", type=("build", "run"))
+    depends_on("py-typing-compat", when="^python@:3.7", type=("build", "run"))
+
+    # A typo in the 0.15.1 setup.py specifies sympy <= 1.9 instead of >= 1.9
```
Can you add a comment pointing to https://github.com/spcl/dace/pull/1483?

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on("py-dill", type=("build", "run"))
+    depends_on("py-fparser@0.1.3:", type=("build", "run"))
+
+    depends_on("py-pyreadline", when="platform=win32", type=("build", "run"))
```
```suggestion
    depends_on("py-pyreadline", when="platform=windows", type=("build", "run"))
```

adamjstewart replied:
@alalazo could we add a test for this? As far as I know, Spack doesn't have an easy way to query all valid platform/arch/os's.

tbennun replied:
Fixed in the latest commit, not resolving the comment so @alalazo can see your comment.

alalazo replied:
> could we add a test for this? 

We can probably add an audit, as for other gotchas in directives.

adamjstewart commented on var/spack/repos/builtin/packages/py-fparser/package.py:
```diff
@@ -23,7 +24,7 @@ class PyFparser(PythonPackage):
     depends_on("py-numpy", type=("build", "run"), when="@:0.0.5")
```
Need to update deps:
```toml
[build-system]
requires = [
    "setuptools >= 42",
    "setuptools_scm[toml] >= 6.2",
    "setuptools_scm_git_archive",
    "wheel >= 0.29.0",
]
```
```cfg
install_requires =
        importlib-metadata; python_version >= "3.6"
        setuptools_scm
setup_requires =
        setuptools_scm
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pyreadline/package.py:
```diff
@@ -0,0 +1,19 @@
+    git = "https://github.com/pyreadline/pyreadline.git"
+    pypi = "pyreadline/pyreadline-2.1.zip"
+
+    version("develop", branch="master")
```
```suggestion
    version("master", branch="master")
```

tbennun committed:
Address review comments

tbennun committed:
Style

tbennun commented:
@adamjstewart Thank you for the thorough review! I think I addressed all the comments now.

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on("py-websockets", type=("build", "run"))
+    depends_on("py-requests", when="@:0.15.1", type=("build", "run"))
+    depends_on("py-flask", when="@:0.15.1", type=("build", "run"))
+    depends_on("py-jinja2", type=("build", "run"))
```
jinja2 is optional and only needed to build the docs, I would remove this

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on("py-setuptools", type="build")
+
+    # Dependencies from setup.py
+    depends_on("py-numpy@1.21:", type=("build", "run"))
```
Personally I wouldn't trust requirements.txt for versions

tbennun replied:
Fixed

adamjstewart commented on var/spack/repos/builtin/packages/py-fparser/package.py:
```diff
@@ -11,19 +11,25 @@ class PyFparser(PythonPackage):
 
+    depends_on("py-setuptools@42:", type="build", when="@0.1:")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm@6.2:+toml", type="build", when="@0.1:")
```
Missing build dep on py-setuptools-scm-git-archive

adamjstewart replied:
Also, setup.cfg claims that setuptools-scm is required at run-time as well

adamjstewart commented on var/spack/repos/builtin/packages/py-fparser/package.py:
```diff
@@ -11,19 +11,25 @@ class PyFparser(PythonPackage):
     depends_on("py-nose", type=("build", "run"), when="@:0.0.7")
     # six is unused as of 0.0.15, but still listed in setup.py
-    depends_on("py-six", type=("build", "run"), when="@0.0.6:")
+    depends_on("py-six", type=("build", "run"), when="@0.0.6:0.1")
```
```suggestion
    depends_on("py-six", type=("build", "run"), when="@0.0.6:0.0")
```
Version ranges are inclusive

tbennun replied:
I'm aware. So 0.1 also covers 0.1.3? What about using 0.1.0 instead here, just to clarify?

tbennun replied:
Just in case, I used 0.0.16, since that is the last release that Spack supports.

adamjstewart replied:
0.1 mean all 0.1.X versions. So both `:0.0` and `:0.0.16` are equivalent.

tbennun committed:
Address review comments

tbennun committed:
Merge branch 'develop' into py-dace

adamjstewart commented on var/spack/repos/builtin/packages/py-dace/package.py:
```diff
@@ -0,0 +1,53 @@
+    # Dependencies from setup.py
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("py-networkx@2.5:", type=("build", "run"))
+    depends_on("py-astunparse@1.6.3:", type=("build", "run"))
```
I wouldn't trust requirements.txt

tbennun replied:
You’re right, I missed it. Thank you!

adamjstewart reviewed (dismissed):
Last comment is optional, let me know if you just want me to merge.

tbennun committed:
Address review comments

tbennun commented:
@adamjstewart  fixed, thank you for noticing! I think it’s ready for merging

adamjstewart commented:
@spackbot run pipeline
